### PR TITLE
Multithreaded Local Search

### PR DESCRIPTION
**Pull request by ge0ffrey**
TODO:
- clean up threads
- manager thread lifecycle in solver
- fix race condition PLANNER-1088

@rsynek Out of about 30 runs, I saw 2 runs hanging, probably due to a race condition.
The first time, it was in a big 3-day benchmark, on the 3 subsingle of SA after 10 subsingles of TS.
The second time, it was on step 600 on a small CB TS configuration, so it's not that it ran out of resources.

https://issues.jboss.org/browse/PLANNER-1088
**I have not been able to reproducible reliably.** We need to first find a way to trigger it more frequently (I suspect playing with the buffer size, moveThreadCount, random sleeps, etc, might do it.)

I now suspect the hanging is due to a dead-lock.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/RearrangingBlockingQueue.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    /**
+     * Not thread-safe. Can only be called from the solver thread.
+     * @return never null
+     * @throws InterruptedException if interrupted
+     * @see BlockingQueue#take()
+     */
+    public MoveResult<Solution_> take() throws InterruptedException {
+        searchMoveIndex++;
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/RearrangingBlockingQueue.java]: @rsynek this causes a one-off issue that deadlocks moveThreadCount=1 with buffer=1, because the searchMoveIndex is initialized to 0, so the first one it's looking for is moveIndex 1, but the first moveIndex is 0.

Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/RearrangingBlockingQueue.java]: The question is, if this is what triggered PLANNER-1088...

Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/RearrangingBlockingQueue.java]: @ge0ffrey truth, this might be just the first step to solve it.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ThreadUtils {
+
+    protected static final transient Logger logger = LoggerFactory.getLogger(ThreadUtils.class);
```
Comment (yurloc) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java]: Why protected? Why transient?

File: optaplanner-docs/src/main/asciidoc/OptimizationAlgorithms/OptimizationAlgorithms-chapter.adoc [hunk 3]
```diff
@@ -759,17 +759,112 @@ To avoid those, configure a `NoChangePhase`:
...
+  <threadFactoryClass>...MyAppServerThreadFactory</threadFactoryClass>
+  ...
+</solver>
+----
+
+The following ``moveThreadCount``s are supported:
+
+* `NONE` (default): Don't run any move threads. Use the single threaded code.
+* ``AUTO``: Let Planner decide how many move threads to run in parallel.
+On machines or containers with little or no CPU's, this can fall back to the single threaded code.
```
Comment (yurloc) [inline optaplanner-docs/src/main/asciidoc/OptimizationAlgorithms/OptimizationAlgorithms-chapter.adoc]: Should be "CPUs" (in this case is's plural).

File: optaplanner-docs/src/main/asciidoc/OptimizationAlgorithms/OptimizationAlgorithms-chapter.adoc [hunk 4]
```diff
@@ -759,17 +759,112 @@ To avoid those, configure a `NoChangePhase`:
...
+----
+
+It is counter-effective to set a moveThreadCount that is higher than the number of available CPU cores,
+as that will slow down the score calculation speed.
+One good reason to do it anyway, is to reproduce a bug of a high-end production machine.
+
+[NOTE]
+====
+Multithreaded solving is _still reproducible_, as long as the resolved `moveThreadCount` is stable.
+A run of the same solver configuration on 2 machines with a different number of CPU's,
```
Comment (yurloc) [inline optaplanner-docs/src/main/asciidoc/OptimizationAlgorithms/OptimizationAlgorithms-chapter.adoc]: CPUs.

File: optaplanner-docs/src/main/asciidoc/PartitionedSearch/PartitionedSearch-chapter.adoc [hunk 5]
```diff
@@ -55,22 +55,21 @@ Advanced configuration:
...
     <localSearch>...</localSearch>
   </partitionedSearch>
 ----
 
-The `threadFactoryClass` allows to plug in a custom `ThreadFactory` for environments
-where arbitrary thread creation should be avoided, such as most application servers, Android, or Google App Engine.
-
 The `runnablePartThreadLimit` allows limiting CPU usage to avoid hanging your machine, see below.
 
+To run in a environment that doesn't like arbitrary thread creation,
```
Comment (yurloc) [inline optaplanner-docs/src/main/asciidoc/PartitionedSearch/PartitionedSearch-chapter.adoc]: "an environment"

File: optaplanner-docs/src/main/asciidoc/PlannerConfiguration/PlannerConfiguration-chapter.adoc [hunk 6]
```diff
@@ -136,7 +136,7 @@ They build the runtime components (of the package namespace ``org.optaplanner.co
 
 [IMPORTANT]
 ====
-The `SolverFactory` is only multi-thread safe after its configured.
+The `SolverFactory` is only thread-safe after its configured.
```
Comment (yurloc) [inline optaplanner-docs/src/main/asciidoc/PlannerConfiguration/PlannerConfiguration-chapter.adoc]: "it's configured"

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                logger.trace("{}        Move index ({}) not doable, ignoring move ({}).",
+                        logIndentation, foragingMoveIndex, foragingMove);
+            } else {
+                moveScope.setScore(result.getScore());
+                boolean accepted = acceptor.isAccepted(moveScope);
+                moveScope.setAccepted(accepted);
+                logger.trace("{}        Move index ({}), score ({}), accepted ({}), move ({}).",
+                        logIndentation,
+                        foragingMoveIndex, moveScope.getScore(), moveScope.getAccepted(),
+                        foragingMove);
```
Comment (ron-murhammer) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Would it be better to rewrap the lines of these logging parameters?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 8]
```diff
@@ -0,0 +1,407 @@
...
+            OrderByMoveIndexBlockingQueue.MoveResult<Solution_> result;
+            try {
+                result = resultQueue.take();
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                break;
+            }
+            if (stepIndex != result.getStepIndex()) {
+                throw new IllegalStateException("Impossible situation: the solverThread's stepIndex (" + stepIndex
+                        + ") differs from the result's stepIndex (" + result.getStepIndex() + ").");
```
Comment (ron-murhammer) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Is this just some kind of sanity check or have you seen this happen? Just seems like a very interesting situation and wonder if its necessary.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 9]
```diff
@@ -0,0 +1,407 @@
...
+                                    logIndentation, moveThreadIndex, stepIndex, moveIndex, score);
+                            // Deliberately add to fail fast if there is not enough capacity (which is impossible)
+                            resultQueue.addMove(moveThreadIndex, stepIndex, moveIndex, move, score);
+                        }
+                    } else {
+                        throw new IllegalStateException("Unknown operation (" + operation + ").");
+                    }
+                    // TODO checkYielding();
+                }
+                logger.trace("{}            Move thread finished.");
```
Comment (ron-murhammer) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Is there a missing log parameter (logIndentation) here?

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+    public static void shutdownAwaitOrKill(ExecutorService executor, String logIndentation, String name) {
+        // Intentionally clearing the interrupted flag so that awaitTermination() in step 3 works.
+        if (Thread.interrupted()) {
+            // 2a. If the current thread is interrupted, propagate interrupt signal to children by initiating
```
Comment (ron-murhammer) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java]: Looks like this code was refactored into this new class from `DefaultPartitionedSearchPhase.java`. Probably should at least remove the numbers in the comments since there is no longer a step 1 otherwise its a big confusing.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java [hunk 11]
```diff
@@ -257,6 +293,37 @@ protected RandomFactory buildRandomFactory(EnvironmentMode environmentMode_) {
         return randomFactory;
     }
 
+    protected Integer resolveMoveThreadCount() {
```
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java]: Could use unit tests

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 12]
```diff
@@ -0,0 +1,407 @@
...
+
+        private int moveThreadIndex;
+        private InnerScoreDirector<Solution_> scoreDirector = null;
+
+        public MoveThreadRunner(int moveThreadIndex) {
+            this.moveThreadIndex = moveThreadIndex;
+        }
+
+        @Override
+        public void run() {
```
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Too long and complicated method, breaking it into private (or package private) methods would make it more readable and testable.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/OrderByMoveIndexBlockingQueue.java [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+import org.optaplanner.core.api.score.Score;
+import org.optaplanner.core.impl.heuristic.move.Move;
+
+public class OrderByMoveIndexBlockingQueue<Solution_> {
```
Comment (rsynek) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/OrderByMoveIndexBlockingQueue.java]: Could use some unit testing

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java (hunk 7 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: What do you mean with rewrap? As in logger.isTrace() ? The getScore() and getAccepted() methods are simple getters, they would be inlined normally.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 14]
```diff
@@ -0,0 +1,407 @@
...
+    protected final ThreadFactory threadFactory;
+    protected final int moveThreadCount;
+    protected final int selectedMoveBufferSize;
+
+    protected boolean assertStepScoreFromScratch = false;
+    protected boolean assertExpectedStepScore = false;
+    protected boolean assertShadowVariablesAreNotStaleAfterStep = false;
+
+    protected List<MoveThreadRunner> moveThreadRunnerList;
+    protected BlockingQueue<MoveThreadOperation<Solution_>> operationQueue;
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Agreed - it needs to be consistent. I've made Operation classes protected, so in case people want to experiment with them. Do note that all these classes are in the impl package, so no BC guarantee.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java (hunk 2 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java]: Agreed, static field loggers should be private and non-transient. I used that form for consistency. So now I was able to consistently fix that :) some benchmark classes adjusted too

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java (hunk 8 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: It's a fail fast. More specifically, because it says "impossible situation", it's a fail fast that won't be triggered by a user doing something wrong (like most fail fasts), but by an intrinsic bug in optaplanner-core itself.
Because the if check is extremely cheap, it's ok to do this.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java (hunk 9 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: Yes, thanks for reporting :)

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java (hunk 10 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/thread/ThreadUtils.java]: Yes, we need to fix that, but I think we might need to simplify the whole thing here. It's probably overkilled.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java [hunk 15]
```diff
@@ -257,6 +293,37 @@ protected RandomFactory buildRandomFactory(EnvironmentMode environmentMode_) {
...
+            }
+        } else {
+            resolvedMoveThreadCount = ConfigUtils.resolveThreadPoolSizeScript(
+                    "moveThreadCount", moveThreadCount, MOVE_THREAD_COUNT_NONE, MOVE_THREAD_COUNT_AUTO);
+        }
+        if (resolvedMoveThreadCount < 1) {
+            throw new IllegalArgumentException("The moveThreadCount (" + moveThreadCount
+                    + ") resulted in a resolvedMoveThreadCount (" + resolvedMoveThreadCount
+                    + ") that is lower than 1.");
+        }
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java]: Yes, it could. Findbugs or PMD might complain about this too.
But we might add other options later (other if else's) and this is a fail-fast check that happens after the number has been resolved. I am going to leave it as-is. Thanks for reporting though.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java [hunk 16]
```diff
@@ -257,6 +293,37 @@ protected RandomFactory buildRandomFactory(EnvironmentMode environmentMode_) {
...
 
+    protected Integer resolveMoveThreadCount() {
+        int availableProcessorCount = Runtime.getRuntime().availableProcessors();
+        Integer resolvedMoveThreadCount;
+        if (moveThreadCount == null || moveThreadCount.equals(MOVE_THREAD_COUNT_NONE)) {
+            return null;
+        } else if (moveThreadCount.equals(MOVE_THREAD_COUNT_AUTO)) {
+            // Leave one for the Operating System and 1 for the solver thread, take the rest
+            resolvedMoveThreadCount = (availableProcessorCount - 2);
+            if (resolvedMoveThreadCount <= 1) {
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java]: 1 move thread != single threaded. In single-threaded there are no move threads. With 1 move thread, there are 2 threads, because there is also the solver thread.

For AUTO, we replace 1 move thread with single threaded. Comment improved to make this clearer:
// To deliberately enforce 1 moveThread, set the moveThreadCount explicitly to 1.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java [hunk 17]
```diff
@@ -136,10 +146,16 @@ public void setPhaseConfigList(List<PhaseConfig> phaseConfigList) {
...
+            threadFactory = ConfigUtils.newInstance(this, "threadFactoryClass", threadFactoryClass);
+        } else {
+            threadFactory = solverConfigPolicy.buildThreadFactory(ChildThreadType.PART_THREAD);
+        }
         DefaultPartitionedSearchPhase phase = new DefaultPartitionedSearchPhase(
                 phaseIndex, solverConfigPolicy.getLogIndentation(), bestSolutionRecaller,
                 buildPhaseTermination(phaseConfigPolicy, solverTermination),
-                buildSolutionPartitioner(), buildThreadFactory(), resolvedActiveThreadCount());
+                buildSolutionPartitioner(), threadFactory, resolvedActiveThreadCount());
         List<PhaseConfig> phaseConfigList_ = phaseConfigList;
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java]: If phaseConfigList is empty, then phaseConfigList_ get defaulted to a CH+LS. But if that would change the global variable, then the benchmarker's report would not contain the exact same solver config in XML form, as the solver config changed when one of the global vars in the config objects change.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/localsearch/LocalSearchPhaseConfig.java [hunk 18]
```diff
@@ -128,15 +131,31 @@ private LocalSearchDecider buildDecider(HeuristicConfigPolicy configPolicy, Term
...
+        if (moveThreadCount == null) {
+            decider = new LocalSearchDecider(configPolicy.getLogIndentation(),
+                    termination, moveSelector, acceptor, forager);
+        } else {
+            Integer moveThreadBufferSize = configPolicy.getMoveThreadBufferSize();
+            if (moveThreadBufferSize == null) {
+                // TODO Verify this is a good default by more meticulous benchmarking on multiple machines and JDK's
+                // If it's too low, move threads will need to wait on the buffer, which hurts performance
+                // If it's too high, more moves are selected that aren't foraged
+                moveThreadBufferSize = 10;
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/localsearch/LocalSearchPhaseConfig.java]: I think that would reduce vertical density
and as that value isn't used anywhere, I don't see any gain to counteract that.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java [hunk 19]
```diff
@@ -0,0 +1,407 @@
...
+        phaseScope.addChildThreadsScoreCalculationCount(childThreadsScoreCalculationCount);
+        moveThreadRunnerList = null;
+        operationQueue = null;
+        resultQueue = null;
+    }
+
+    private ExecutorService createThreadPoolExecutor() {
+        ThreadPoolExecutor threadPoolExecutor
+                = (ThreadPoolExecutor) Executors.newFixedThreadPool(moveThreadCount, threadFactory);
+        if (threadPoolExecutor.getMaximumPoolSize() < moveThreadCount) {
```
[the remark this answers is not in the document]
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/localsearch/decider/multithreaded/MultiThreadedLocalSearchDecider.java]: I have no idea. The JDK javadocs were ambiguous.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java (hunk 11 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java]: Yes, let's make a jira around this? I 'll add some unit tests, but probably after summit (merging won't be immediate anyway).

**Comment by ge0ffrey** [comment]
jenkins retest this

File: optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java]: thanks

File: optaplanner-core/src/main/java/org/optaplanner/core/config/localsearch/LocalSearchPhaseConfig.java (hunk 18 above)
[the remark this answers is not in the document]
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/localsearch/LocalSearchPhaseConfig.java]: that's a valid point, I agree

File: optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java (hunk 11 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/solver/SolverConfig.java]: https://issues.jboss.org/browse/PLANNER-1099

**Comment by ge0ffrey** [comment]
jenkins retest this with the latest master

File: optaplanner-core/src/test/java/org/optaplanner/core/impl/solver/thread/OrderByMoveIndexBlockingQueueTest.java [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+import org.optaplanner.core.impl.testdata.domain.TestdataSolution;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.junit.Assert.assertEquals;
+import static org.optaplanner.core.impl.testdata.util.PlannerAssert.*;
+
+public class OrderByMoveIndexBlockingQueueTest {
+
+    private static final Logger logger = LoggerFactory.getLogger(PartitionQueueTest.class);
```
Comment (yurloc) [inline optaplanner-core/src/test/java/org/optaplanner/core/impl/solver/thread/OrderByMoveIndexBlockingQueueTest.java]: Should be `OrderByMoveIndexBlockingQueueTest.class`.

**Comment by ge0ffrey** [comment]
@rsynek By writing OrderByMoveIndexBlockingQueueTest, I also thought by adding the test addExceptionIsNotEatenIfNextStepStartsBeforeTaken() - and guess what - there was an edge case that exceptions got eaten... Fixing now (and committing test then too)

**Comment by ge0ffrey** [comment]
@rsynek I did a bunch more changes, to also enable it for Construction Heuristics. I'd like to merge this PR soon (early next week).
